bsr_dense_mm(): better test coverage

nikitaved opened the pull request:
This PR improves test coverage for `bsr_dense_mm` by:
- ~~enabling correctness tests for `float32`~~.
- extending and testing input correctness checks.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100882
* #100876
* __->__ #100543



cc @alexsamardzic @pearu @cpuhrsch @amjames @bhosmer

nikitaved committed:
bsr_dense_mm(): better test coverage

[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- enabling correctness tests for `float32`
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- enabling correctness tests for `float32`
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- enabling correctness tests for `float32`
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- enabling correctness tests for `float32`
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- enabling correctness tests for `float32`
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- enabling correctness tests for `float32`
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- enabling correctness tests for `float32`
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- enabling correctness tests for `float32`
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- enabling correctness tests for `float32`
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved commented:
@cpuhrsch , @malfet , that single failure is very strange. When I log in into the CI and re-run, all the tests are green there.

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- enabling correctness tests for `float32`
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- enabling correctness tests for `float32`
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- enabling correctness tests for `float32`
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- enabling correctness tests for `float32`
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- enabling correctness tests for `float32`
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- ~~enabling correctness tests for `float32`~~.
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

malfet commented on test/test_sparse_csr.py:
```diff
@@ -3397,6 +3398,51 @@ def test_triton_bsr_dense_bmm(self, device, dtype, index_dtype, block_size):
 
+    @onlyCUDA
+    @skipIfRocm
+    @dtypes(torch.half)
+    @dtypesIfCUDA(torch.half)
```
One of the two is redundant, isn't it?
```suggestion
    @dtypes(torch.half)
```

nikitaved replied:
It throws `AssertionError: @dtypes is mandatory when using @dtypesIf however 'test_triton_bsr_dense_bmm_error_messages' didn't specify it`

malfet commented on test/test_sparse_csr.py:
```diff
@@ -3397,6 +3398,51 @@ def test_triton_bsr_dense_bmm(self, device, dtype, index_dtype, block_size):
+            bsr_dense_mm(lhs.to_sparse_bsc(16), rhs)
+        with self.assertRaisesRegex(ValueError, "on the same GPU device"):
+            bsr_dense_mm(lhs, rhs.cpu())
+        if torch.cuda.device_count() > 1:
+            with self.assertRaisesRegex(ValueError, "on the same GPU device"):
+                bsr_dense_mm(lhs.to("cuda:0"), rhs.to("cuda:1"))
```
We should keep all mutli-GPU tests separate (and decorated with `@requiresMultiGPU` or something like that)

malfet commented on torch/sparse/_triton_ops.py:
```diff
@@ -463,19 +460,42 @@ def check(cond, msg):
             )
 
+            def is_power_of_two(v):
+                return not (v & (v - 1))
```
Nit (I'm not sure how `not` is defined for integers in Python, perhaps you can send a link
```suggestion
                return (v & (v - 1)) == 0
```

nikitaved replied:
https://docs.python.org/3/reference/expressions.html#not

malfet replied:
My point here, is that, in my mind,  mixing logical `and` and boolean `not` hurts readability,  on the other hand, it's  up to an author/maintainer of the codebase. I.e. as long as  @cpuhrsch is fine reading that, I'm ok as well.

malfet commented on torch/sparse/_triton_ops.py:
```diff
@@ -463,19 +460,42 @@ def check(cond, msg):
-            row_block = bsr.values().shape[-2]
+            row_block, col_block = bsr.values().shape[-2:]
             check(
                 not n % row_block,
```
Nit (same as below, not sure about `not` definition for Python and operator precedence, wheres `x % y == z` is much more straightforward, isn't it?
```suggestion
               n % row_block == 0,
```

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- ~~enabling correctness tests for `float32`~~.
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- ~~enabling correctness tests for `float32`~~.
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved committed:
Update on "bsr_dense_mm(): better test coverage"

This PR improves test coverage for `bsr_dense_mm` by:
- ~~enabling correctness tests for `float32`~~.
- extending and testing input correctness checks.




cc alexsamardzic pearu cpuhrsch amjames bhosmer

[ghstack-poisoned]

nikitaved commented:
@pytorchbot merge
